### PR TITLE
Chore/admin semantic table rows

### PR DESCRIPTION
**Pull request by MadelineCollier**
## Summary

This PR is for issue: [[Admin] Table component rows do not contain links](https://github.com/solidusio/solidus/issues/5944)

This addresses the major concerns of that issue, but does not remove the click handler as it's still required in one case. That is the case of the `<tr>` clicks. I tried inserting an `<a>` to wrap the whole row, but that didn't work. The browser refused to render it as it seemingly violates the HTML standard for tables. You can have a link, a td, but not both on the same level within the tr. You can only have both if the link is nested inside the td. We might want to look into inserting it inside of each of the `<tr>`'s `<td>`s, but this seems quite messy since some `<td>`s currently have their own behaviour  which would be overruled by that link (row checkbox inputs for one example).

So to recap, this PR addresses our over reliance on the `rowClicked` stimulus handler for all cases except for the `<tr>` itself. Now instead of rendering text or divs, we render semantic UI elements (links) with proper hrefs. Hopefully this makes sense @tvdeyen let me know what you think.

## Before:

https://github.com/user-attachments/assets/3d676a32-a8d2-4f5b-96f2-4bc4c70c9439

## After:

https://github.com/user-attachments/assets/004a9975-4597-4dbe-bf31-3d590fdf20a1

## Checklist

Check out our [PR guidelines](https://github.com/solidusio/.github/blob/master/CONTRIBUTING.md#pull-request-guidelines) for more details.

The following are mandatory for all PRs:

- [x] I have written a thorough PR description.
- [x] I have kept my commits small and atomic.
- [x] [I have used clear, explanatory commit messages](https://github.com/solidusio/.github/blob/main/CONTRIBUTING.md#writing-good-commit-messages).

The following are not always needed:

- 📖 I have updated the README to account for my changes.
- 📑 I have documented new code [with YARD](https://www.rubydoc.info/gems/yard/file/docs/Tags.md).
- 🛣️ I have opened a PR to update the [guides](https://github.com/solidusio/edgeguides).
- ✅ I have added automated tests to cover my changes.
- 📸 I have attached screenshots to demo visual changes.

### CODE REVIEW

**Review by kennyadsl — COMMENTED** [review]
Nice! What about using `link_to` instead?

File: admin/app/components/solidus_admin/ui/table/component.html.erb [hunk 1]
```diff
@@ -146,7 +146,7 @@
           <% end %>
         >
           <% @data.columns.each do |column| %>
-            <%= render_data_cell(column, row) %>
+            <%= render_data_cell(column, row, @data.url&.call(row)) %>
```
Comment (tvdeyen) [inline admin/app/components/solidus_admin/ui/table/component.html.erb:149]: Can you explain what this change does? I can try to figure out, but I am not yet very fluent in this part of Solidus yet.

File: admin/app/components/solidus_admin/users/index/component.rb [hunk 2]
```diff
@@ -83,15 +83,21 @@ def columns
       },
       {
         header: :order_count,
-        data: ->(user) { user.order_count },
+        data: ->(user) do
+          content_tag :a, user.order_count, href: row_url(user)
```
Comment (tvdeyen) [inline admin/app/components/solidus_admin/users/index/component.rb:87]: `link_to` seems to be the helper we want here, no?

File: admin/app/components/solidus_admin/users/index/component.rb [hunk 3]
```diff
@@ -83,15 +83,21 @@ def columns
...
+        end
       },
       {
         header: :lifetime_value,
-        data: -> { _1.display_lifetime_value.to_html },
+        data: ->(user) do
+          content_tag :a, user.display_lifetime_value.to_html, href: row_url(user)
```
Comment (tvdeyen) [inline admin/app/components/solidus_admin/users/index/component.rb:93]: ditto

File: admin/app/components/solidus_admin/users/index/component.rb [hunk 4]
```diff
@@ -83,15 +83,21 @@ def columns
...
+        end
       },
       {
         header: :last_active,
-        data: ->(user) { last_login(user) },
+        data: ->(user) do
+          content_tag :a, last_login(user), href: row_url(user)
```
Comment (tvdeyen) [inline admin/app/components/solidus_admin/users/index/component.rb:99]: ditto

**Review by tvdeyen — COMMENTED** [review]
Alberto is right, We should use the `link_to` helper wherever possible.

Regarding the `tr` click event: I consider this a harmful and bad UX pattern and we should remove it completely.

**Comment by MadelineCollier** [comment]
For sure, I can switch out `content_tag` for `link_to`! @tvdeyen what do suggest with regards to my comments in the PR description regarding the issues with removing the click handler for the `<tr>` elements. We definitely still need these table rows to be clickable as that is essentially the backbone of the new admin flow. Table rows either redirect to individual edit pages or open modals. I was not able to find a way to replace the click handler with a link as I commented above. What else do you suggest or do you think I have missed in my tests?

**Comment by tvdeyen** [comment]
> We definitely still need these table rows to be clickable as that is essentially the backbone of the new admin flow. 

Technically we do not need table rows to be clickable in order for the new admin flow to work. Clicking links inside the data cells should open modals or links, not clicking the row itself. It is not possible to wrap a TR in a A in HTML for good reasons.

Consider a table row having multiple "actions" inside (let's say the promotions table has a link to edit the promotion itself and a promotion code, if clicked, visits the promotion codes list). How do we make this possible if only one click action is available for a whole row?

> Table rows either redirect to individual edit pages or open modals. I was not able to find a way to replace the click handler with a link as I commented above. What else do you suggest or do you think I have missed in my tests?

Thanks to your work we now have links in all table rows. So we should have the same UX as before, but on the HTML elements that are accessibly and semantically the correct elements. The click events inside the cells should bubble to the handler, so the click handler should still do its work. Ideally the row event handler would only handle modals and nothing else, but this is something to tackle later.

Sorry for all the push back. I know that neither you nor I have been involved in designing or developing the new admin. We now have the burden to make it as maintainable as possible while still getting it done.

I appreciate all the work you put into it, but please bear with we that I am looking at it from a maintainers prospective.

Thanks 🙏🏻

**Comment by tvdeyen** [comment]
All of these are solvable by using less code and more semantics :)

1. We should make linked text inside of the table row visually stand out. Currently there is extra CSS code that makes the links not appear as links. This is bad for a11y and easily be solved by letting the browser do what a browser does: Visually indicate the this is an actionable item (usually by underlining the text or by giving it different color)
2. Not all components need do be clickable. It does not make sense to click a "complete" pill and assume the edit order link to open. It would make much more sense to filter the table by this state instead. But it is totally ok that a table cell just contains text. It is a list of information and some of the information can be clicked. 
3. We could add a link on all "name"/"number"/"primary indicator" columns, if. it already the case. 

I think we should not overcomplicate things. The browser, HTML and CSS provide us with simple, semantic, established and accessible tools that all do not require more, but less code.

The web can be reduced to two simple UX patterns:  links and forms. It's that simple.

**Comment by MadelineCollier** [comment]
That sounds reasonable @tvdeyen I can take a look at this today. I'll do a pass over all the primary identifiers (most of which should already be links), swap out the content_tag for link_to, and address the spec failures and we should be good.

File: promotions/lib/components/admin/solidus_promotions/promotion_categories/index/component.rb [hunk 5]
```diff
@@ -40,7 +40,7 @@ def name_column
     {
       header: :name,
       data: ->(promotion_category) do
-        content_tag :div, promotion_category.name
+        content_tag :a, promotion_category.name, href: row_url(promotion_category)
```
Comment (kennyadsl) [inline promotions/lib/components/admin/solidus_promotions/promotion_categories/index/component.rb:43]: You probably missed this one. 🙂

File: promotions/lib/components/admin/solidus_promotions/promotion_categories/index/component.rb [hunk 6]
```diff
@@ -49,7 +49,7 @@ def code_column
     {
       header: :code,
       data: ->(promotion_category) do
-        content_tag :div, promotion_category.code
+        content_tag :a, promotion_category.code, href: row_url(promotion_category)
```
Comment (kennyadsl) [inline promotions/lib/components/admin/solidus_promotions/promotion_categories/index/component.rb:52]: and this one

File: legacy_promotions/lib/components/admin/solidus_admin/promotion_categories/index/component.rb [hunk 7]
```diff
@@ -40,7 +40,7 @@ def name_column
     {
       header: :name,
       data: ->(promotion_category) do
-        content_tag :div, promotion_category.name
+        content_tag :a, promotion_category.name, href: row_url(promotion_category)
```
Comment (kennyadsl) [inline legacy_promotions/lib/components/admin/solidus_admin/promotion_categories/index/component.rb:43]: and this!

**Review by kennyadsl — COMMENTED** [review]
Thansk @MadelineCollier, I marked some other occurrences of `content_tag :a`, let's please fix them, unless there's a reason to keep them as they are.

File: legacy_promotions/lib/components/admin/solidus_admin/promotion_categories/index/component.rb (hunk 7 above)
Comment (MadelineCollier) [reply legacy_promotions/lib/components/admin/solidus_admin/promotion_categories/index/component.rb]: thanks!

File: promotions/lib/components/admin/solidus_promotions/promotion_categories/index/component.rb (hunk 5 above)
Comment (MadelineCollier) [reply promotions/lib/components/admin/solidus_promotions/promotion_categories/index/component.rb]: fixed!

File: promotions/lib/components/admin/solidus_promotions/promotion_categories/index/component.rb (hunk 6 above)
Comment (MadelineCollier) [reply promotions/lib/components/admin/solidus_promotions/promotion_categories/index/component.rb]: fixed!

**Comment by MadelineCollier** [comment]
Thanks for catching those @kennyadsl, fixed! Still having some trouble with the specs. It seems like now that the rowClicked handler isn't in charge anymore, the links aren't preserving the query params (and many specs assert that they should be preserved even when opening/closing modals). Any ideas on how to fix this behaviour? We won't know what the ransack search params are at time of render as far as I can tell. I can also remove those spec assertions if we think its reasonable that if you open a modal to edit a table row, the query params are allowed to be lost.

**Comment by kennyadsl** [comment]
I can't recall exactly what was the need for that requirement. Is there a PR linked to that failing specs that we can reach with a git blame?

**Comment by MadelineCollier** [comment]
> I can't recall exactly what was the need for that requirement. Is there a PR linked to that failing specs that we can reach with a git blame?

No blames that I found were particularly helpful as the specs were generally just added as part of the development of the new admin. It makes sense that where possible we’d want to retain the filters/searches that users have inputted, but I can find a good way of doing that for all of our new links. This largely only comes into play on edit actions triggered by table link clicks.

**Comment by kennyadsl** [comment]
Yes, you are right. It was probably because when we search for something, edit an item on the row and close the modal, we want to keep the existing selection of the table. cc'ing @elia and @rainerdema to see if we have more details here. 

I think we should keep this behavior, as the alternative is a very bad UX where you need to filter again if you want to edit multiple items in the same table after a specific filtering.

**Comment by MadelineCollier** [comment]
Ok guys, I have been plugging away at the specs, and while the consistent blocker is the missing query params, another issue has come up that relates to the `rowClicked` handler.

See it in it's original form:
```
  rowClicked(event) {
    // If the user clicked on a link, button, input or summary, skip the row url visit
    if (event.target.closest("td").contains(event.target.closest("a,select,textarea,button,input,summary"))) return

    if (this.modeValue === "batch") {
      this.toggleCheckbox(event.currentTarget)
    } else {
      const url = new URL(event.params.url, "http://dummy.com")
      const params = new URLSearchParams(url.search)
      const frameId = params.get('_turbo_frame')
      const frame = frameId ? { frame: frameId } : {}
      // remove the custom _turbo_frame param from url search:
      params.delete('_turbo_frame')
      url.search = params.toString()

      window.Turbo.visit(url.pathname + url.search, frame)
    }
  }
```

The deletion of the `_turbo_frame` param turns out to be important for some of our admin pages. Not doing this results in  some of our admin edit modals succesfully submitting, updating the given model, but **not closing the modal**. So the turbostream refresh happens, but the target is still the edit modal, and the page is re-rendered with it open. We could redirect back to the index page but this would clear the query params, and also kind of defeat the purpose of using turboframe modals to begin with. [There is discussion around the ability to override the target from the server response here](https://github.com/hotwired/turbo/issues/257), but nothing solid yet that I have seen in examples.

At this point it feels like 1 step forward, 3 steps back. It would be nice to hear from @elia and @rainerdema to see if this is a path (more semantic UI links instead of stimulus handlers) that has already been explored. This would explain why the `rowClicked` handler exists today, as it looks like it currently serves two purposes that plain links can't (as far as I have seen in my tests). @tvdeyen I can definitely still appreciate the maintenance angle for the original issue/request to change things up, but hesitate to spend another week on this while so much of the admin remains unfinished and my availability on the project winds down. @kennyadsl Also curious to hear your thoughts on the pros/cons of this.

**Comment by tvdeyen** [comment]
A tricky problem we faced in Alchemy as well. It has been solved by responding with a proper 422 on errors, so that the modal gets redisplayed and properly closed via a redirect from the server. Query parameters are preserved by passing around the ransack parameters. 

All that said, I know that this is a tough problem to solve and I was under the impression it can be solved in the new admin. 

Unfortunately there is much more involved to fix this issue.

I can totally understand if you want to table this for now. 

Maybe I find some time in the future to contribute some of the code we built for Alchemy to Solidus. 

Sorry for these circumstances

**Comment by MadelineCollier** [comment]
Ok, in that case, with your blessing @tvdeyen I’ll table this PR for now and get back to the other admin work. Hopefully the learnings from this PR and comment thread can inform the issue whenever we choose to pick it up in the future.

**Comment by tvdeyen** [comment]
Hey @MadelineCollier can we try to just extract 6bd7f5e14d460687420208ec96ccfde98da29f38 from this PR? So that we have the links in place? It should work just fine with the row click event handler.

**Comment by MadelineCollier** [comment]
> Hey @MadelineCollier can we try to just extract [6bd7f5e](https://github.com/solidusio/solidus/commit/6bd7f5e14d460687420208ec96ccfde98da29f38) from this PR? So that we have the links in place? It should work just fine with the row click event handler.

The links don't work with the specs as they blow away the query params, something which the specs expect to be present after clicking through. Since we still haven't solved that I think we'll need to keep it as is until we have a solution to preserve those params.

**Comment by tvdeyen** [comment]
Ok, I figured it out. A combination of `target="_top"` on the `turbo_frame_tag` of the modals form and a proper redirect instead of the `turbo_frame` refresh in the controllers works as expected. Actually what one would do in regular CRUD apps as well. I will prepare a PR soon.

**Comment by tvdeyen** [comment]
@MadelineCollier This is my current draft PR https://github.com/solidusio/solidus/pull/6014 
Love to have your feedback

**Pull request closed without merging**